pints/_mcmc/_nuts.py

```diff
@@ -0,0 +1,719 @@
+
+    # find a good value to start epsilon at
+    # (this will later be refined so that the acceptance probability matches
+    # delta)
+    epsilon = yield from find_reasonable_epsilon(theta, L, grad_L, step_size)
+
+    # default values taken from [1]
+    mu = np.log(10 * epsilon)
+    log_epsilon_bar = np.log(1)
+    H_bar = 0
+    gamma = 0.05
+    t0 = 10
+    kappa = 0.75
+
+    # start at iteration 1
+    m = 1
+
+    # provide an infinite generator of mcmc steps....
+    while True:
+        # randomly sample momentum
+        r0 = np.random.normal(size=len(theta))
+        hamiltonian0 = L - 0.5 * r0.dot(r0)
+
+        if use_multinomial_sampling:
+            # use multinomial sampling
+            log_u = hamiltonian0
+            # n is a tree weight using a float
+            n = 0.0
+        else:
+            # use slice sampling
+            log_u = np.log(np.random.uniform(0, 1)) + hamiltonian0
+            # n is the number of accepted points in the chain
+            n = 1
+
+        # create initial integration path state
+        state = NutsState(theta, r0, L, grad_L, n, 1, None, None, False,
+                use_multinomial_sampling)
+        j = 0
+
+        # build up an integration path with 2^j points, stopping when we either
+        # encounter a U-Turn, or reach a max number of points 2^10
+        while j < max_tree_depth and state.s == 1:
+
+            # pick a random direction to integrate in
+            # (to maintain detailed balance)
+            if np.random.randint(0, 2):
+                vj = 1
+            else:
+                vj = -1
+
+            # recursivly build up tree in that direction
+            state_dash = yield from \
+                build_tree(state, log_u, vj, j, epsilon,
+                           hamiltonian0, step_size, hamiltonian_threshold,
+                           use_multinomial_sampling)
+            state.update(state_dash, direction=vj, root=True)
+
+            j += 1
+
+        # adapt epsilon using dual averaging
+        if m < M_adapt:
+            H_bar = (1 - 1.0 / (m + t0)) * H_bar + 1.0 / (m + t0) * \
+                (delta - state.alpha / state.n_alpha)
+            log_epsilon = mu - (np.sqrt(m) / gamma) * H_bar
+            log_epsilon_bar = m**(-kappa) * log_epsilon + \
+                (1 - m**(-kappa)) * log_epsilon_bar
+            epsilon = np.exp(log_epsilon)
+        elif m == M_adapt:
+            epsilon = np.exp(log_epsilon_bar)
+
+        # update current position in chain
+        theta = state.theta
+        L = state.L
+        grad_L = state.grad_L
+
+        # signal calling process that mcmc step is complete by passing a tuple
+        # (rather than an ndarray)
+        yield (theta, L, state.alpha / state.n_alpha, 2**j, state.divergent)
+
+        # next step
+        m += 1
+
+
+class NoUTurnMCMC(pints.SingleChainMCMC):
+    r"""
+
+    Implements No U-Turn Sampler (NUTS) with dual averaging, as described in
+    Algorithm 6 in [1]_.
```

Comment:
Cool. Can we add a note to say that it either does Algo 6 in [1] or does multinomial sampling. In the latter case, can we include pseudocode or something (seeing as it's not written anywhere in Stan!).

Reply:
where should the pseudocode go? In the docstring, can this be formatted nicely?